From 64caed614eaeb781a7c74062a029276545e4a20c Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Fri, 7 Jul 2023 09:03:25 +0600
Subject: [PATCH 71/92] intswcd: fix wrong WARN_ON() firing

Core interrupt handling code can invoke chip's irq_ack() method for
level interrupts, thus must not fire WARN_ON() in that case.

[This is a lost change caused by importing wrong branch as vc4-on-3.13,
on 3.14 port it shall be amended into "irqchip: renesas INTSWCD driver"
commit]

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 drivers/irqchip/irq-renesas-intswcd.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/irqchip/irq-renesas-intswcd.c b/drivers/irqchip/irq-renesas-intswcd.c
index 76c9933151e2..d2161788e9d1 100644
--- a/drivers/irqchip/irq-renesas-intswcd.c
+++ b/drivers/irqchip/irq-renesas-intswcd.c
@@ -76,10 +76,9 @@ static void intswcd_irq_ack(struct irq_data *d)
 	struct intswcd *swcd = d->domain->host_data;
 	const struct intswcd_hwirq_info *ihi = &swcd->ii->ihi[d->hwirq];
 
-	if (WARN_ON_ONCE(ihi->type != INTSWCD_EDGE))
-		return;
-
-	iowrite32(ihi->mask, edge_clr_reg(swcd, ihi->reg));
+	/* This is called both for level and for edge irqs */
+	if (ihi->type == INTSWCD_EDGE)
+		iowrite32(ihi->mask, edge_clr_reg(swcd, ihi->reg));
 }
 
 static struct irq_chip intswcd_irq_chip = {
-- 
2.30.2

